### PR TITLE
Add rewriters chain loading by configuration (backwards compatibility) #76

### PR DESCRIPTION
**Pull request by tkaessmann**
As part of the #76 it's now possible to still parse the deprecated rewriters configuration (instead of the new REST interface). If you want to use this, you have to move the rewriters configuration to a listener in the solrconfig like:

```xml
<listener event="firstSearcher" class="querqy.solr.ClassicRewriteChainLoader">  
   <lst name="rewriteChain">
      <lst name="rewriter">
      ...
      </lst>
   </lst>
</listener>
```
And set the `inMemory` configuration in the new request handler:
```xml
<requestHandler name="/querqy/rewriter" class="querqy.solr.QuerqyRewriterRequestHandler">
   <bool name="inMemory">true</bool>
</requestHandler>
```
thanks also to @tboeghk

### CODE REVIEW

**Review by renekrie — CHANGES_REQUESTED** [review]

**Comment by tboeghk** [comment]
@tkaessmann note to self: Let's merge the master into this PR to resolve future merge and testing hell :-/

File: querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/ClassicRewriteChainLoader.java [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+                        final String className = (String) config.get("class");
+
+                        final Map<String, Object> jsonBody = ((ClassicConfigurationParser) SolrRewriterFactoryAdapter
+                                .loadInstance(id, className))
+                                .parseConfigurationToRequestHandlerBody((NamedList<Object>) config, resourceLoader);
+
+                        final SolrQueryRequestBase req = new SolrQueryRequestBase(core, params) {
+                        };
+                        req.setContentStreams(newArrayList(new ContentStreamBase.StringStream(JsonUtil.toJson(jsonBody), UTF_8.name())));
```
Comment (renekrie) [inline querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/ClassicRewriteChainLoader.java:98]: newArrayList - this list can be immutable. We can simply use https://docs.oracle.com/javase/6/docs/api/java/util/Collections.html#singletonList(T) and get rid of the Guava dependency at the same go.

File: querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/QuerqyQParserPlugin.java [hunk 2]
```diff
@@ -68,6 +69,12 @@ public void init(final @SuppressWarnings("rawtypes") NamedList args) {
             }
         }
 
+        // This element is not allowed any more here and must be removed
+        if (args.get("rewriters") != null) {
+            throw new IllegalArgumentException("'rewriters' configuration is not allowed anymore in the query parser configuration. " +
+                    "You have to move it to a Querqy listener or change the rewriter deployment. See upgrade info.");
```
Comment (renekrie) [inline querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/QuerqyQParserPlugin.java:75]: Let's mention the concrete Listener class in the error message

File: querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/QuerqyQParserPlugin.java [hunk 3]
```diff
@@ -269,4 +267,18 @@ public final QParser createParser(final String qstr, final SolrParams localParam
...
+
+    private QuerqyRewriterRequestHandler getQuerqyRequestHandler(final SolrCore core){
+        final SolrRequestHandler requestHandler = core.getRequestHandler(rewriterRequestHandlerName);
+        if (requestHandler == null) {
+            throw new SolrException(SolrException.ErrorCode.SERVER_ERROR,
+                    "Could not find QuerqyRewriterRequestHandler for name " + rewriterRequestHandlerName);
+        }
+        if (!(requestHandler instanceof QuerqyRewriterRequestHandler)) {
+            throw new SolrException(SolrException.ErrorCode.SERVER_ERROR,
+                    rewriterRequestHandlerName + " is not a " + QuerqyRewriterRequestHandler.class);
```
Comment (renekrie) [inline querqy-for-lucene/querqy-solr/src/main/java/querqy/solr/QuerqyQParserPlugin.java:279]: QuerqyRewriterRequestHandler.class.getName() ...

File: querqy-for-lucene/querqy-solr/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker [hunk 4]
```diff
@@ -0,0 +1 @@
+mock-maker-inline
```
Comment (renekrie) [inline querqy-for-lucene/querqy-solr/src/test/resources/mockito-extensions/org.mockito.plugins.MockMaker:1]: Wow! I didn't know that we can mock final classes!

**Review by renekrie — APPROVED** [review]
Please take note of the comments. I'm going to approve the PR and resolve a few minor things myself.

**Comment by renekrie** [comment]
Thank you @tkaessmann  & @tboeghk

**Pull request merged**